src/ros/lowbandwidth_trajectory_follower.cpp

```diff
@@ -0,0 +1,386 @@
+        l_stopping = g_stopping
+        exit_critical
+        while not l_stopping:
+            while l_sent_waypoints_number == l_requested_waypoints_number and not l_stopping:
+                sleep(SERVOJ_TIME_WAITING)
+
+                enter_critical
+                l_requested_waypoints_number = g_requested_waypoints_number
+                l_stopping = g_stopping
+                exit_critical
+
+            end
+            if l_stopping:
+                break
+            end
+            send_message(l_sent_waypoints_number + 1)
+            l_sent_waypoints_number = l_sent_waypoints_number + 1
+        end
+        join controlling_thread
+    end
+
+    thread receivingThread():
+        local sending_thread = run sendingThread()
+        while True:
+            waypoint_received = socket_read_ascii_float(14, CONNECTION_NAME)
+            if waypoint_received[0] == 0:
+                textmsg("Not received trajectory for the last 2 seconds. Quitting")
+                enter_critical
+                g_stopping = True
+                exit_critical
+                break
+            elif waypoint_received[0] != JOINT_NUM * 2 + 2:
+                textmsg("Received wrong number of floats in trajectory. This is certainly not OK.")
+                textmsg(waypoint_received[0])
+                enter_critical
+                g_stopping = True
+                exit_critical
+                break
+            elif is_waypoint_sentinel(waypoint_received):
+                add_next_waypoint(waypoint_received)
+                enter_critical
+                g_stopping = True
+                g_received_waypoints_number = g_received_waypoints_number + 1
+                exit_critical
+                break
+            end
+            add_next_waypoint(waypoint_received)
+        end
+        join sending_thread
+    end
+    socket_open(REVERSE_IP, REVERSE_PORT, CONNECTION_NAME)
+    receiving_thread = run receivingThread()
+    join receiving_thread
+    socket_close(CONNECTION_NAME)
+    textmsg("Exiting the program")
+end
+
+)";
+
+LowBandwidthTrajectoryFollower::LowBandwidthTrajectoryFollower(URCommander &commander, std::string &reverse_ip, int reverse_port,
+                                       bool version_3)
+  : running_(false)
+  , commander_(commander)
+  , server_(reverse_port)
+  , time_interval_(0.008)
+  , servoj_time_(0.008)
+  , servoj_time_waiting_(0.001)
+  , max_waiting_time_(2.0)
+  , servoj_gain_(300.0)
+  , servoj_lookahead_time_(0.03)
+  , max_joint_difference_(0.01)
+{
+  ros::param::get("~time_interval", time_interval_);
+  ros::param::get("~servoj_time", servoj_time_);
+  ros::param::get("~servoj_time_waiting", servoj_time_waiting_);
+  ros::param::get("~max_waiting_time", max_waiting_time_);
+  ros::param::get("~servoj_gain", servoj_gain_);
+  ros::param::get("~servoj_lookahead_time", servoj_lookahead_time_);
+  ros::param::get("~max_joint_difference", max_joint_difference_);
```

Comment:
Could you please pass these in as parameters after parsing them in the `ros_main.cpp:parse_args` function? 
Especially because stuff like servoj_gain and such should be passed the same way to the regular `TrajectoryFollower` (seems I never got around to fixing that either)

Reply:
I think it is kind of nice to keep those in here. Rosparam is actually a nice abstraction that allows to read the params exactly in the place where they are needed rather than reading them in a "central" param reading place and passing those as parameters. I rather like the approach that parameters are parsed close to the place where they are used. That follows more the "distribution" and "decentralisation" pattern (more like micro-services") which is quite centerpiece of ROS's distributed nodes approach (but on internal node level). 

The good thing about it is that when you read the code, you can exactly see which parameters are "global" - used more for the whole driver and which are "local" - used only by some components (like the lowbandwidth_follow_joint_trajectory). I think it quite makes sense to keep them here. Do you think there is a benefit in moving them to ros_main ?

Reply:
The primary reason I have for moving them to `ros_main` is keeping the ROS dependency to a minimum such that people could easily take the code and use it directly from C++ without involving ROS. 
But perhaps that's a bit too idealistic since I don't know of anyone actually using it that way so I guess your justification is sufficient to do it your way.